redisson/src/main/java/org/redisson/RedissonPermitExpirableSemaphore.java

```diff
@@ -295,114 +305,158 @@ private CompletableFuture<String> acquireAsync(int permits, RedissonLockEntry en
-        return commandExecutor.syncedEval(getRawName(), ByteArrayCodec.INSTANCE, RedisCommands.EVAL_PERMIT_DATA,
-                  "local expiredIds = redis.call('zrangebyscore', KEYS[2], 0, ARGV[4], 'limit', 0, ARGV[1]); " +
+    private RFuture<List<String>> tryAcquireAsync(List<String> ids, long timeoutDate) {
+        List<Object> params = new ArrayList<>();
+        params.add(ids.size());
+        params.add(timeoutDate);
+        params.add(System.currentTimeMillis());
+        params.add(nonExpirableTimeout);
+        params.add(getSubscribeService().getPublishCommand());
+
+        for (String permitId: ids) {
+            params.add(ByteBufUtil.decodeHexDump(permitId));
+        }
+        
+        CompletionStage<List<String>> future = commandExecutor.syncedEval(getRawName(), ByteArrayCodec.INSTANCE, RedisCommands.EVAL_STRING,
+                  "local expiredIds = redis.call('zrangebyscore', KEYS[2], 0, ARGV[3], 'limit', 0, ARGV[1]); " +
                   "if #expiredIds > 0 then " +
                       "redis.call('zrem', KEYS[2], unpack(expiredIds)); " +
                       "local value = redis.call('incrby', KEYS[1], #expiredIds); " +
                       "if tonumber(value) > 0 then " +
-                          "redis.call(ARGV[6], KEYS[3], value); " +
+                          "redis.call(ARGV[5], KEYS[3], value); " +
                       "end;" +
                   "end; " +
                   "local value = redis.call('get', KEYS[1]); " +
                   "if (value ~= false and tonumber(value) >= tonumber(ARGV[1])) then " +
                       "redis.call('decrby', KEYS[1], ARGV[1]); " +
-                      "redis.call('zadd', KEYS[2], ARGV[2], ARGV[3]); " +
+                      "for i = 6, #ARGV, 1 do " +
+                          "redis.call('zadd', KEYS[2], ARGV[2], ARGV[i]); " +
+                      "end; " +
 
                       "local ttl = redis.call('pttl', KEYS[1]); " +
                       "if ttl > 0 then " +
                           "redis.call('pexpire', KEYS[2], ttl); " +
                       "end; " +
 
-                      "return ARGV[3]; " +
+                      "return 'OK'; " +
                   "end; " +
                   "local v = redis.call('zrange', KEYS[2], 0, 0, 'WITHSCORES'); " +
-                  "if v[1] ~= nil and v[2] ~= ARGV[5] then " +
+                  "if v[1] ~= nil and v[2] ~= ARGV[4] then " +
                       "return ':' .. tostring(v[2]); " +
                   "end " +
                   "return nil;",
                   Arrays.asList(getRawName(), timeoutName, channelName),
-                permits, timeoutDate, id, System.currentTimeMillis(), nonExpirableTimeout, getSubscribeService().getPublishCommand());
+                  params.toArray()
+        ).thenApply(result -> {
+            if (result == null) {
+                return Collections.emptyList();
+            }
+            if (result.equals("OK")) {
```

Comment:
Please use a boolean result instead

Reply:
I thought about it, but I also need to return closest timeout [here](https://github.com/redisson/redisson/pull/5247/files/d513705bc6dd6cfa8dc76c04e18bba74d45783e1#diff-a6afb637de0f8d40a841263def0f289d6200822972dde04dfed78baf9a77239cR410)